### PR TITLE
Add OpenShift Client extension

### LINKED ISSUE #3200: Implement OpenshiftClient on Kubernetes Client Extension
**Description**
I've been using Kubernetes Client Extension for a while and I can't find a way to create a `DefaultOpenShiftClient` [1]. It seems to me that this class is not available yet on Quarkus 0.19.0. Am I right?

Any plans to solve this? 

[1] https://github.com/fabric8io/kubernetes-client

### PR DESCRIPTION
**Pull request by vsevel**
This extension provides access to OpenShift clusters through the [fabric8 openshift client](https://github.com/fabric8io/kubernetes-client/tree/master/openshift-client).
It relies heavily on the existing [quarkus kubernetes client extension](https://quarkus.io/guides/kubernetes-client) for its implementation (e.g. it shares the same configuration).

Essentially it brings an `OpenShiftClient` that can be injected into the client application.

Fixes https://github.com/quarkusio/quarkus/issues/3200

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/Feature.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/Feature.java:57]: Please, alphabetical order :).

File: extensions/openshift-client/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gsmet) [inline extensions/openshift-client/runtime/src/main/resources/META-INF/quarkus-extension.yaml:6]: If we don't have a guide, let's drop this and add it once the guide is there.

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (gsmet) [inline extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java:129]: Shouldn't part of this be shared to avoid having to fix bugs in both extension? Most of this doesn't look specific to OpenShift?

File: extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java
Comment (gsmet) [inline extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java:12]: Let's make it `@Singleton` maybe?

**Comment by vsevel** [comment]
hello @geoand  here is a first draft. several things need to be discussed or reworked.

In `OpenShiftClientProcessor` I basically copied everything from `KubernetesClientProcessor`, and changed the `Feature.KUBERNETES_CLIENT` to be `OPENSHIFT_CLIENT`, and commented out the `roleProducer.produce...`.
I am not sure how much I needed from that class. if we need to keep most of it, we would need to refactor it in a place where both extensions could reuse it, instead of copying it in the openshift client extension.

I did not support the `openshift-server-mock`, but I suppose we should do something like was done for the `kubernetes-server-mock` in the `test-framework`?

I did not work on the regression testing. I put some code in to play with an external openshift server, but I left the tests disabled for now. 

However, I did a full native test based on the `OpenShiftClientTestApp` class. So I was able to produce a native image and execute it against a live openshift cluster. 

Would you be able to give me an initial review?

File: integration-tests/openshift-client/src/test/java/io/quarkus/openshift/client/runtime/OpenShiftClientTestIT.java
Comment (gsmet) [inline integration-tests/openshift-client/src/test/java/io/quarkus/openshift/client/runtime/OpenShiftClientTestIT.java:6]: Why is it disabled?

File: integration-tests/openshift-client/src/main/resources/application.properties
Comment (gsmet) [inline integration-tests/openshift-client/src/main/resources/application.properties:5]: I wonder if we want to keep the `kubernetes-client` prefix for the config but it's just a question.

**Review by gsmet — DISMISSED** [review]
Thanks for this!

I haven't done a full review as I'm not very familiar with these things but I added a few suggestions and questions.

For the questions, probably better to wait for someone more familiar with it to step up.

**Comment by geoand** [comment]
Thanks a lot!

I'll do a thorough review, but probably not until Thursday.

Also including @iocanel and @gastaldi in case they want to take a look

**Comment by gsmet** [comment]
@vsevel I have another question: AFAICS, you include the Kubernetes Client deployment artifact so I'm wondering why you need to duplicate things?

**Comment by vsevel** [comment]
> you include the Kubernetes Client deployment artifact so I'm wondering why you need to duplicate things?

@gsmet this is really fresh paint, and knowingly unfinished work. I did not know how much logic I needed to get from `KubernetesClientProcessor` (specifically I am lacking knowledge on the use cases it addresses, and how much code we need to reuse). But yes we will need to refactor the code to avoid duplication.

> I wonder if we want to keep the kubernetes-client prefix for the config but it's just a question.

my understanding is that the configuration is shared between the fabric8 `KubernetesClient `and `OpenShiftClient`. so I was wondering if we should just honor that and keep the exact same properties for the extensions. it is definitely up for discussion.

> Let's make it @Singleton maybe?

I just copied it from the kubernetes client extension. If `Singleton` is better, then we should correct it also there.

> Why is it disabled?

this is really sandbox code at this point. I am waiting for an answer regarding the mock server before I start implementing the regression tests.

File: integration-tests/openshift-client/src/main/resources/application.properties
Comment (gastaldi) [reply integration-tests/openshift-client/src/main/resources/application.properties:5]: It would be nice if we could specify aliases for the `ConfigRoot` annotations

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (gastaldi) [inline extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java:28]: This seems to have been copied from `KubernetesClientProcessor`. Since this extension depends on kubernetes-client, this is not needed. It should contain only what's necessary for `OpenShiftClient`

**Review by gastaldi — CHANGES_REQUESTED** [review]
The OpenShiftProcessor needs to be reviewed, since a lot of code is already in the KubernetesProcessor

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (gastaldi) [inline extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java:30]: The kubernetes client extension already does that

Comment (vsevel) [reply extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java:30]: it does this:
```
featureProducer.produce(new FeatureBuildItem(Feature.KUBERNETES_CLIENT));
sslNativeSupport.produce(new ExtensionSslNativeSupportBuildItem(Feature.KUBERNETES_CLIENT));
```

not this:
```
featureProducer.produce(new FeatureBuildItem(Feature.OPENSHIFT_CLIENT));
sslNativeSupport.produce(new ExtensionSslNativeSupportBuildItem(Feature.OPENSHIFT_CLIENT));
```

don't we want the feature to be displayed?
and for ssl support I agree.
it was not clear to me whether or not things needed to be repeated in the context of the new extension (I am not entirely comfortable with processors).

File: extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java
Comment (vsevel) [reply extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java]: by the way, documentations provides un example with `ApplicationScoped`
should it be changed as well?

File: integration-tests/openshift-client/src/test/java/io/quarkus/openshift/client/runtime/OpenShiftClientTestIT.java
Comment (vsevel) [reply integration-tests/openshift-client/src/test/java/io/quarkus/openshift/client/runtime/OpenShiftClientTestIT.java]: I have implemented the tests.

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (gastaldi) [reply extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java]: Yes, I was referring to the SSL support. Thanks!

**Comment by iocanel** [comment]
@vsevel: Thank you so much for it, I really needed that.

File: extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/OpenShiftClientProducer.java
Comment (iocanel) [inline extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/OpenShiftClientProducer.java:17]: `openshiftClient` maybe?

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (iocanel) [reply extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java]: If there is a way it would be nice to use it, as more client extensions could be added in the future. Personally I'd love to see `kantive` and `tekton`. So that would be awesome!

File: integration-tests/openshift-client/src/main/resources/application.properties
Comment (iocanel) [reply integration-tests/openshift-client/src/main/resources/application.properties]: The Config object is shared between `kubernetes-client` and `openshfit-client` so I am not really sure if it makes sense to use a different prefix. After all, they are all `kuberrnetes`.

**Review by iocanel — DISMISSED** [review]
Looks good. Apart from a minor comment that is optional really, I would only like to see if possible sharing the native configuration stuff for jackson, okhttp etc (if possible).

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (gastaldi) [inline extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java:24]: No big deal, but you can move the `featureProducer.produce(...)` method to a separate method in the processor, like in the example below

https://github.com/quarkusio/quarkus/blob/770dbae2a93876bee33d52fb3c275c5549bc6aab/extensions/jgit/deployment/src/main/java/io/quarkus/jgit/deployment/JGitProcessor.java#L16-L19

File: extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientBuildConfig.java
Comment (gastaldi) [inline extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientBuildConfig.java:95]: Nice! :+1:

Comment (vsevel) [reply extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientBuildConfig.java:95]: thanks. yes that was missing for my tests ;)

**Comment by gastaldi** [comment]
In Fabric8 Launcher we used to have an impersonation feature, maybe that would be nice to have too? Could be in a separate PR/issue:

https://github.com/fabric8-launcher/launcher-application/blob/master/services/openshift-service-impl/src/main/java/io/fabric8/launcher/service/openshift/impl/Fabric8OpenShiftServiceImpl.java#L118-L124

**Comment by vsevel** [comment]
when I build a native image, I get a lot of thoses warnings:
`[WARNING] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Duplicate entry ValidationMessages.properties entry from io.fabric8:openshift-model::jar:4.10.3(compile) will be ignored. Existing file was provided by io.fabric8:kubernetes-model-core::jar:4.10.3(compile)`
and other files such as schema/validation-schema.json, schema/kube-validation-schema.json, ...
I don't know if this is something we should concerned about, and if there is anything we could do.

**Comment by gastaldi** [comment]
> on windows, when I build a native image, I get a lot of thoses warnings:
> `[WARNING] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Duplicate entry ValidationMessages.properties entry from io.fabric8:openshift-model::jar:4.10.3(compile) will be ignored. Existing file was provided by io.fabric8:kubernetes-model-core::jar:4.10.3(compile)`
> and other files such as schema/validation-schema.json, schema/kube-validation-schema.json, ...
> I don't know if this is something we should concerned about, and if there is anything we could do.

@vsevel open an issue in https://github.com/fabric8io/kubernetes-client/issues, this seems to likely be a build issue

**Comment by gastaldi** [comment]
@vsevel in fact this file is a leftover from the Bean Validation integration that was removed long ago. I'll create a PR removing it.

**Comment by gastaldi** [comment]
PR provided: https://github.com/fabric8io/kubernetes-client/pull/2473

**Comment by gastaldi** [comment]
About the `schema/*.json` files I am not sure how they are used, maybe @iocanel or @gsmet may know more?

**Comment by vsevel** [comment]
I think it is also a consequence of:
`[WARNING] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Uber JAR strategy is used for native image source JAR generation on Windows. This is done for the time being to work around a current GraalVM limitation on Windows concerning the maximum command length (see https://github.com/oracle/graal/issues/2387).`

**Comment by gastaldi** [comment]
@vsevel right, but regardless of the chosen strategy I think these files in the k8s client build need to be reviewed anyway

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (geoand) [reply extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java]: +1 for what both Guillaume and Ioannis said above

File: extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java
Comment (geoand) [reply extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java]: Yeah, it should be `@Singleton`

File: integration-tests/openshift-client/src/main/resources/application.properties
Comment (geoand) [reply integration-tests/openshift-client/src/main/resources/application.properties]: +1 for what Ioannis said

File: extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java
Comment (vsevel) [reply extensions/openshift-client/deployment/src/main/java/io/quarkus/openshift/client/deployment/OpenShiftClientProcessor.java]: > Shouldn't part of this be shared to avoid having to fix bugs in both extension? Most of this doesn't look specific to OpenShift?

this code is not duplicated anymore.
are you suggesting to create a base kubernetes extension that includes part of the `KubernetesClientProcessor`? then, it would have to depend on the fabric8 kubernetes-client library, right; because the referred classes such as `io.fabric8.kubernetes.api.model.IntOrString` come from that library. 
there are not `knative` or `tekton` libraries today? if there were, you idea is to get them to share those common classes? then we are talking about some code reuse in the fabric8 project with support for `knative` and `tekton` modules, before even talking about some code reuse in `quarkus`?
what exactly do you think we can do at this point here in `quarkus`?

File: extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java
Comment (vsevel) [reply extensions/openshift-client/runtime/src/main/java/io/quarkus/openshift/client/runtime/OpenShiftClientProducer.java]: corrected in doc "https://quarkus.io/guides/writing-extensions"
see https://github.com/quarkusio/quarkus/pull/11982/commits/4e5a98a611343bebeaa19628c067f06fb3b3bb05

**Comment by vsevel** [comment]
> In Fabric8 Launcher we used to have an impersonation feature, maybe that would be nice to have too? Could be in a separate PR/issue:

@geoand only the `Impersonate-User` attribute? (i.e. not `Impersonate-Group` and `Impersonate-Extra-( extra name )`?)
cf https://kubernetes.io/docs/reference/access-authn-authz/authentication/

**Comment by geoand** [comment]
> > In Fabric8 Launcher we used to have an impersonation feature, maybe that would be nice to have too? Could be in a separate PR/issue:
> 
> @geoand only the `Impersonate-User` attribute? (i.e. not `Impersonate-Group` and `Impersonate-Extra-( extra name )`?)
> cf https://kubernetes.io/docs/reference/access-authn-authz/authentication/

@gastaldi ^

**Comment by gastaldi** [comment]
Ah didn't know it was part of Kubernetes already. I think it makes sense to have that implemented in the `kubernetes-client` extension. Let's do it in a separate issue/PR

**Comment by vsevel** [comment]
I have added some minimalistic documentation as a section in the original `kubernetes-client` extension itself: https://github.com/quarkusio/quarkus/pull/11982/commits/23afa0f66836ed9b744d252fe5e5737b86e79235
It should be sufficient, since apart from the name changes, usage is identical.

**Comment by vsevel** [comment]
hello @gastaldi 
I have done some additional testing in jvm and native mode. overall this works well.
In my use case, I assume the master url is not known at build time, and will be provided by an `application.properties` mounted in the container with:
```
quarkus.kubernetes-client.master-url=https://...:8443/
```
In the jvm mode I execute:
```
docker run -i --rm -v E:\ocpclient\config:/deployments/config quarkus/ocpclient-jvm
```
and I get:
```
2020-09-15 15:01:58,442 WARN  [io.qua.run.ConfigChangeRecorder] (main) Build time property cannot be changed at runtime. quarkus.kubernetes-client.master-url was null at build time and is now https://...:8443/
...
2020-09-15 15:05:50,978 INFO  [com.lod.art.ocp.PrivilegedProducer] (main) baseConfig => https://...:8443/
```
With the following code:
```
    @Inject
    Config baseConfig;

    @Privileged
    @Produces
    @Singleton
    public OpenShiftClient privileged() {

        log.info("baseConfig => " + baseConfig.getMasterUrl());
```
So there is a warning, but I get the appropriate value.

however, when I execute in native mode, I do get the warning, but the value is wrong (I get the default value):
```
2020-09-15 15:08:00,454 WARN  [io.qua.run.ConfigChangeRecorder] (main) Build time property cannot be changed at runtime. quarkus.kubernetes-client.master-url was null at build time and is now https://...:8443/
...
2020-09-15 15:08:00,458 INFO  [com.lod.art.ocp.PrivilegedProducer] (main) baseConfig => https://kubernetes.default.svc/
```

is it expected to work in jvm mode?
if so it is misleading to display the warning.

is it expected to work in native mode?
if not, then I would argue that the master-url should be a runtime property, not build time, otherwise that would mean we have to produce an executable that is tied to a specific environment. and you could argue the same for some other properties, such as the ones regarding the certs, the auth, ...

in the end we can bypass this by redefining our own properties and overriding the creation. For instance here by providing the master url:
```
@SuppressWarnings("CdiInjectInspection")
@Singleton
public class PrivilegedProducer {

    private static final Logger log = LoggerFactory.getLogger(PrivilegedProducer.class);

    @Inject
    Config baseConfig;

    @ConfigProperty(name = "master-url")
    String masterUrl;

    @Privileged
    @Produces
    @Singleton
    public OpenShiftClient privileged() {

        log.info("baseConfig => " + baseConfig.getMasterUrl());

        Config config = new ConfigBuilder(baseConfig)
                .withMasterUrl(masterUrl)
                ...
                .withTrustCerts(true)
                .build();

        return new DefaultOpenShiftClient(config);
    }

}
```

with the `application.properties`:
```
master-url=https://...:8443/
```

what do you think?

**Comment by gastaldi** [comment]
I think `quarkus.kubernetes-client.master-url` should be a runtime property. WDYT @iocanel @geoand ?

**Comment by geoand** [comment]
> I think `quarkus.kubernetes-client.master-url` should be a runtime property. WDYT @iocanel @geoand ?

Ideally yes, but I don't think it's that critical.

Not only that, but I think we had to make it `BUILD_AND_RUN_TIME_FIXED` in order to make the deployment to Kubernetes work properly

**Comment by vsevel** [comment]
> but I don't think it's that critical.

no it is not, but you could be doing all your dev thinking it is working, and having to introduce your own properties with your own producer when at the end you want to produce a native image. right? anything I am missing?

**Comment by geoand** [comment]
Yeah, I understand. 

But given how unlikely it is for the `masterURL` to be set and the fact that I am pretty sure that moving it to runtime will cause issues with the `quarkus.kubernetes.deploy=true`, I would much rather prefer leaving it as is.

**Comment by vsevel** [comment]
ok. then I believe the extension is ready.
any unresolved points in the conversation or others you think we should address?

**Comment by gastaldi** [comment]
@vsevel please squash your commits before marking it as ready for review. Thanks!

**Comment by vsevel** [comment]
should we have `@PreDestroy` callbacks to be able to close the kubernetes or openshift clients?
I see we have this pattern for numerous produceurs already (e.g. `DynamodbClientProducer`, `S3ClientProducer`, ...)

**Comment by gsmet** [comment]
@vsevel yes, the clients should be closed at `@PreDestroy` if the lifecycle is managed by CDI.

**Comment by vsevel** [comment]
hi @iocanel @gastaldi you still have some unclosed "requested changes". 
I believe everything has been addressed. 
I have added the `PreDestroy`. let me know if something is missing or not, then I will do a final squash.
thanks

**Review by gastaldi — APPROVED** [review]
LGTM

**Comment by vsevel** [comment]
@gsmet I have squashed and rebased. I believe it is ready to be merged, unless you have some more comments? thanks.

**Comment by vsevel** [comment]
@gsmet @gastaldi @iocanel hello all, are we still considering this for 1.9? thanks.

**Comment by gastaldi** [comment]
@vsevel one thing I missed are the changes in the build pipeline to run these extension tests?

**Comment by vsevel** [comment]
> @vsevel one thing I missed are the changes in the build pipeline to run these integration tests?

@gastaldi hmm we are talking about that sort of thing https://github.com/quarkusio/quarkus/blob/master/integration-tests/neo4j/pom.xml#L123-L181 ?

it is true I started the extention from https://github.com/quarkusio/quarkus/blob/master/integration-tests/kubernetes-client/pom.xml, where I do not see it.

**Comment by gastaldi** [comment]
@vsevel No, I was referring to https://github.com/quarkusio/quarkus/blob/master/.github/workflows/ci-actions.yml#L546-L549 specifically

**Comment by vsevel** [comment]
@gastaldi ok so I change it to:
```
          - category: Misc3
            timeout: 65
            test-modules: >
              kubernetes-client
              openshift-client
```
?

**Comment by gastaldi** [comment]
Yeah, I think that makes sense. 

Is `65` a random chosen number or something you measured previously?

**Comment by vsevel** [comment]
> Is `65` a randomly chosen number or something you measured previously?

@gastaldi well it was 60 before, and I remember guillaume saying that we count 5 mins per native test. or it is not part of the native timeout, and I can leave 60?

**Comment by gastaldi** [comment]
No, 65 is fine, go ahead ;)

**Comment by vsevel** [comment]
@gastaldi done. waiting for ci.

File: extensions/kubernetes-client/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientProducer.java
Comment (gastaldi) [inline extensions/kubernetes-client/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientProducer.java:15]: No biggie, but instead of storing it in an attribute to destroy later, you can pass it as a method argument.

Eg. 
```java
@PreDestroy
public void destroy(KubernetesClient client) { 
   client.close();
}
```

Comment (gastaldi) [reply extensions/kubernetes-client/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientProducer.java:15]: Ah ignore me. I am confusing it with CDI's `@Destroyed` (https://docs.jboss.org/cdi/api/2.0/javax/enterprise/context/Destroyed.html)

Comment (vsevel) [reply extensions/kubernetes-client/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientProducer.java]: looks better, and avoids testing for null.

File: extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/runtime/OpenShiftClientProducer.java
Comment (gastaldi) [inline extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/runtime/OpenShiftClientProducer.java:23]: This needs to be 
```java
 public void destroy(@Observes @Destroyed OpenShiftClient client) {
...
}
```

AFAIK

**Review by gastaldi — CHANGES_REQUESTED** [review]
Can you confirm the @PreDestroy changes work as intended? I believe you need to use `@Destroyed` but I'd like some feedback before merging this PR ;)

File: extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/runtime/OpenShiftClientProducer.java
Comment (vsevel) [reply extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/runtime/OpenShiftClientProducer.java:23]: I could not get it to work. `@Destroyed` requires a scope.
I could not find something that looked like what you are suggesting.
do you have another idea, or should I revert the code?

Comment (vsevel) [reply extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/runtime/OpenShiftClientProducer.java]: this works:
```
    public void destroy(@Observes @Destroyed(ApplicationScoped.class) Object container, OpenShiftClient client) {
        if (client != null) {
            LOGGER.info("closing openshift client");
            client.close();
        }
    }
```

I did not know if the client could be null, so I implemented defensively. 
I squashed the commits.

**Comment by vsevel** [comment]
> Can you confirm the @PreDestroy changes work as intended? I believe you need to use `@Destroyed` but I'd like some feedback before merging this PR ;)

yes it does. but you have to observe the destruction of the application scoped context.
waiting for the ci.

**Comment by vsevel** [comment]
@gastaldi I thought I had it, but with the `public void destroy(@Observes @Destroyed(ApplicationScoped.class) Object container, OpenShiftClient client)` approach I am getting this failure on `OpenShiftClientTest`.

```
[INFO] Running io.quarkus.it.openshift.client.OpenShiftClientTest
[ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 5.719 s <<< FAILURE! - in io.quarkus.it.openshift.client.OpenShiftClientTest
[ERROR] io.quarkus.it.openshift.client.OpenShiftClientTest.createRoute  Time elapsed: 0.016 s  <<< ERROR!
java.lang.RuntimeException: 
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.AmbiguousResolutionException: Ambiguous dependencies for type io.fabric8.kubernetes.client.KubernetesClient and qualifiers [@Default]
        - java member: io.quarkus.kubernetes.client.runtime.KubernetesClientProducer#destroy()
        - declared on ObserverInfo [beanClass=io.quarkus.kubernetes.client.runtime.KubernetesClientProducer, priority=2500, isAsync=false, reception=ALWAYS, transactionPhase=IN_PROGRESS, observedType=java.lang.Object, qualifiers=[@Destroyed(value = javax.enterprise.context.ApplicationScoped)]]
        - available beans:
                - PRODUCER METHOD bean [types=[java.lang.AutoCloseable, io.fabric8.kubernetes.client.KubernetesClient, java.lang.Object, java.io.Closeable, io.fabric8.openshift.client.OpenShiftClient, io.fabric8.kubernetes.client.ConfigAware, io.fabric8.kubernetes.client.Client], qualifiers=[@Default, @Any], target=io.fabric8.openshift.client.OpenShiftClient openshiftClient(io.fabric8.kubernetes.client.Config config), declaringBean=io.quarkus.it.openshift.client.runtime.OpenShiftClientProducer]
                - PRODUCER METHOD bean [types=[java.lang.AutoCloseable, io.fabric8.kubernetes.client.KubernetesClient, java.lang.Object, java.io.Closeable, io.fabric8.kubernetes.client.ConfigAware, io.fabric8.kubernetes.client.Client], qualifiers=[@Default, @Any], target=io.fabric8.kubernetes.client.KubernetesClient kubernetesClient(io.fabric8.kubernetes.client.Config config), declaringBean=io.quarkus.kubernetes.client.runtime.KubernetesClientProducer]
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1018)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:238)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:122)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:411)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:936)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

I reverted the code back to the original `@PreDestroy` approach.

**Review by gastaldi — APPROVED** [review]

**Comment by geoand** [comment]
Please rebase this onto the latest master as the Kubernetes Client version was bumped in the meantime

**Comment by vsevel** [comment]
> Please rebase this onto the latest master as the Kubernetes Client version was bumped in the meantime

@gastaldi FYI ci passed with k8s client 4.12.0

**Comment by gastaldi** [comment]
@vsevel thank you!

**Comment by vsevel** [comment]
@gastaldi thanks for the merge

**Pull request merged**
